book/html.md

```diff
@@ -5,408 +5,559 @@ prev: text
-so real HTML parsers try to guess what the author meant and somehow
-return a tree anyway.[^3]
+Use the following `SELF_CLOSING_TAGS` list, straight from the
+[standard][html5-void-elements]:[^void-elements]
 
-For example, some HTML tags serve as both an open and a close tag (they
-are called self-closing tags). The `<br>` tag, which inserts a line
-break, is one such tag. However, right now, our parser doesn\'t know
-that, and will just keep looking for a `</br>` tag that never
-arrives.[^4]
+[html5-void-elements]: https://html.spec.whatwg.org/multipage/syntax.html#void-elements
 
-To support self-closing tags, we just need to modify the part of the
-parser that creates a new node, which reads:
+[^void-elements]: A lot of these tags are obscure or obsolete, but
+    it's nice that there's a complete list.
 
 ``` {.python}
-new = ElementNode(current, tok.tag)
-if current is not None: current.children.append(new)
-current = new
+SELF_CLOSING_TAGS = [
+    "area", "base", "br", "col", "embed", "hr", "img", "input",
+    "link", "meta", "param", "source", "track", "wbr",
+]
 ```
 
-Here, the first line creates the element and the second makes it a child
-of its parent. Then, the third one \"enters\" the new node, in effect
-opening it. For a self-closing tag, we just need to avoid doing that:
+Test your parser on this page to see if that helped.
 
-``` {.python}
-if new.tag != "br":
-    current = new
-```
+::: {.further}
+Some people put a slash at the end of a self-closing tag (like
+`<br/>`) but unlike [XML][xml-self-closing] that final slash is
+totally ignored.
+:::
 
-Now is a good time, by the way, to implement `<br>` in `layout`; it
-works pretty much the same way that `<p>` does, ending the current line
-by resetting `x` and incrementing `y`.
+[xml-self-closing]: https://www.w3.org/TR/xml/#sec-starttags
 
-Self-closing tags aren\'t supposed to have close tags; but sometimes
-people forget close tags for elements that really should have them. For
-example, you might have a `<p>` inside a `<section>` and then close the
-`</section>` without closing the `</p>`. Because these errors are so
-common, browsers try to automatically fix them so they do the \"right
-thing\". The full algorithm is pretty complicated (there are multiple
-types of tags and so on) but let\'s implement a simple version of it:
-when a close tag is encountered, we will walk up the tree until we find
-a tag that it closes.
+Attributes
+==========
+
+It doesn't help; why? Because the problem tags aren't just `<meta>`:
+they have attributes, like in `<meta charset="utf-8" />`. HTML
+attributes give additional information about an element; open tags can
+have any number of attributes (though close tags can't have any).
+Attribute values can be either quoted, unquoted, or omitted entirely.
+
+For simplicity, let's stick to unquoted attribute values. Since
+neither tag names nor attribute-value pairs can contain
+whitespace,[^because-unquoted] we can split the tag contents on
```

Comment:
The contents of an attribute value can contain spaces. Example: the sandbox attribute.
https://html.spec.whatwg.org/#sandboxing

Reply:
You could say instead that for simplicity we are going to ignore attribute values that allow spaces, because those require handling quoted text.

Reply:
Yeah, it's for simplicity. In an earlier draft I pushed a 7-state parsing automaton to handle quoted values, but it was just too long. I'll clarify that.